src/main/java/com/google/cloud/tools/ProcessCaller.java

```diff
@@ -0,0 +1,174 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.tools;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Strings;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.logging.Logger;
+
+/**
+ * Calls external tools like gcloud or dev_appserver. Hides system process invocation logic from
+ * {@link Action}.
+ *
+ * <p>All the logic for generating and running commands is contained in this class.
+ */
+public class ProcessCaller {
+
+  // TODO(joaomartins): Will this work in e.g., Windows?
+  private static File DEFAULT_WORKING_DIR = new File("/");
+  private static Logger LOG = Logger.getLogger(ProcessCaller.class.getName());
+  private List<String> command;
+  private File workingDirectory;
+  private boolean synchronous;
+  private String cloudSdkLocation = System.getProperty("user.home") + "/google-cloud-sdk";
+
+  public ProcessCaller(Tool tool, List<String> arguments, Map<Option, String> flags) {
+    this(tool, arguments, flags, DEFAULT_WORKING_DIR, true);
+  }
+
+  public ProcessCaller(Tool tool, List<String> arguments, Map<Option, String> flags,
+      File workingDirectory) {
+    this(tool, arguments, flags, workingDirectory, true);
+  }
+
+  public ProcessCaller(Tool tool, List<String> arguments, Map<Option, String> flags,
+      boolean synchronous) {
+    this(tool, arguments, flags, DEFAULT_WORKING_DIR, synchronous);
+  }
+
+  public ProcessCaller(Tool tool, List<String> arguments, Map<Option, String> flags,
+      File workingDirectory, boolean synchronous) {
+    this.workingDirectory = workingDirectory;
+    this.synchronous = synchronous;
+    this.command = prepareCommand(tool, arguments, flags);
+  }
+
+  public boolean call() throws GCloudErrorException {
+    ProcessBuilder processBuilder = new ProcessBuilder(command);
+    processBuilder.directory(workingDirectory);
+    processBuilder.inheritIO();
+    Process gcloudProcess;
+    try {
+      gcloudProcess = processBuilder.start();
+    } catch (IOException ioEx) {
+      LOG.severe(String.format("Error running gcloud CLI. %s", ioEx));
+      return false;
+    }
+
+    // If call is asynchronous, don't wait for the process to end.
+    if (synchronous) {
+      try {
+        int exitStatus = gcloudProcess.waitFor();
+
+        if (exitStatus != 0) {
+          throw new GCloudErrorException(exitStatus);
+        }
+      } catch (InterruptedException intEx) {
+        LOG.severe(String.format("Process interrupted. Reason %s", intEx));
+      }
+    }
+
+    return true;
+  }
+
+  /**
+   * Finds the executable path for gcloud.
+   */
+  @VisibleForTesting
+  public String getGcloudPath() {
+    String gcloudLocation = cloudSdkLocation + "/bin/gcloud";
```

Comment:
cloudSdkLocation is set by default to System.getProperty("user.home") + "/google-cloud-sdk". I still need to test in Windows.

The reason I'm not converting to File is because I actually just need the string location, to pass to ProcessBuilder. I would have to encode into a File, then decode, having no other use for the File object.

Something I don't think is going to work in Windows is

private static File DEFAULT_WORKING_DIR = new File("/");